awx/main/scheduler/task_manager_models.py

```diff
@@ -64,61 +137,77 @@ def __contains__(self, hostname):
         self.instance_groups = dict()
+        self.task_manager_instances = task_manager_instances if task_manager_instances is not None else TaskManagerInstances()
         self.controlplane_ig = None
         self.pk_ig_map = dict()
+        self.control_task_impact = kwargs.get('control_task_impact', settings.AWX_CONTROL_NODE_TASK_IMPACT)
+        self.controlplane_ig_name = kwargs.get('controlplane_ig_name', settings.DEFAULT_CONTROL_PLANE_QUEUE_NAME)
 
         if instance_groups is not None:  # for testing
-            self.instance_groups = instance_groups
+            self.instance_groups = {ig.name: TaskManagerInstanceGroup(ig, self.task_manager_instances, **kwargs) for ig in instance_groups}
+            self.pk_ig_map = {ig.pk: ig for ig in instance_groups}
         else:
             if instance_groups_queryset is None:
-                instance_groups_queryset = InstanceGroup.objects.prefetch_related('instances').only('name', 'instances')
+                instance_groups_queryset = InstanceGroup.objects.prefetch_related('instances').only(
+                    'name', 'instances', 'max_concurrent_jobs', 'max_forks', 'is_container_group'
+                )
             for instance_group in instance_groups_queryset:
-                if instance_group.name == settings.DEFAULT_CONTROL_PLANE_QUEUE_NAME:
+                if instance_group.name == self.controlplane_ig_name:
                     self.controlplane_ig = instance_group
-                self.instance_groups[instance_group.name] = dict(
-                    instances=[
-                        instances_by_hostname[instance.hostname] for instance in instance_group.instances.all() if instance.hostname in instances_by_hostname
-                    ],
-                )
+                self.instance_groups[instance_group.name] = TaskManagerInstanceGroup(instance_group, self.task_manager_instances, **kwargs)
                 self.pk_ig_map[instance_group.pk] = instance_group
 
+    def __getitem__(self, ig_name):
+        return self.instance_groups.get(ig_name)
+
+    def __contains__(self, ig_name):
+        return ig_name in self.instance_groups
+
     def get_remaining_capacity(self, group_name):
-        instances = self.instance_groups[group_name]['instances']
-        return sum(inst.remaining_capacity for inst in instances)
+        return self.instance_groups[group_name].get_remaining_instance_capacity()
 
     def get_consumed_capacity(self, group_name):
-        instances = self.instance_groups[group_name]['instances']
-        return sum(inst.consumed_capacity for inst in instances)
+        return self.instance_groups[group_name].get_consumed_instance_capacity()
+
+    def consume_capacity(self, task):
+        self.task_manager_instances.consume_capacity(task)
+        if task.instance_group_id:
+            ig = self.pk_ig_map[task.instance_group_id]
+            if ig.is_container_group:
+                self.instance_groups[ig.name].consume_capacity(task)
```

Comment:
I find it a little unexpected that `TaskManagerInstanceGroups.consume_capacity` deducts the capacity from the instance group as well as the instance(s) the job runs on.

There are 4 methods for `consume_capacity`, and this isn't a bad thing, but with that number it becomes hard to predict, so it might be best to follow expectations as closely as possible. As it stands, it might be more consistent to introduce a `TaskManagerModels.consume_capacity` method to deduct capacity from both the IGs and the instances, moving the `self.task_manager_instances.consume_capacity(task)` line from here to there. Maybe ask @fosterseth for his input on this matter as well. My suggestion here would require changing the task manager to keep the `TaskManagerModels` instance as opposed to the `TaskManagerInstanceGroups`, but that might obligate some other changes to the references between classes.

Reply:
I agree its a bit weird to have the method on `TaskManagerInstanceGroups` but have it apply to groups and instances, I'll shop some ideas around with seth and we can move forward. I don't think any other place other than the task manager ever uses any `consume_capacity` function so it should be relatively small lift

Reply:
implemented the suggestion